core/src/subgraph/instance_manager.rs

```diff
@@ -456,83 +451,101 @@ where
-
-        let created_ds_count = block_state.created_data_sources.len();
-        let block_with_calls = EthereumBlockWithCalls {
-            ethereum_block: block.deref().clone(),
-            calls,
-        };
-
-        future::result(<B>::Stream::parse_triggers(
-            EthereumLogFilter::from_data_sources(data_sources.iter()),
-            EthereumCallFilter::from_data_sources(data_sources.iter()),
-            EthereumBlockFilter::from_data_sources(data_sources.iter()),
-            false,
-            block_with_calls,
-        ))
-        .from_err()
-        .and_then(move |block_with_triggers| {
-            let triggers = block_with_triggers.triggers;
+    .and_then(move |(ctx, block_state)| {
+        // If new data sources have been created, restart the subgraph after this block.
+        let needs_restart = !block_state.created_data_sources.is_empty();
+
+        // This loop will:
+        // 1. Instantiate created data sources.
+        // 2. Process those data sources for the current block.
+        // Until no data sources are created or MAX_DATA_SOURCES is hit.
+
+        // Note that this algorithm processes data sources spawned on the same block _breadth
+        // first_ on the tree implied by the parent-child relationship between data sources. Only a
+        // very contrived subgraph would be able to observe this.
+        loop_fn(
+            (ctx, block_state),
+            move |(mut ctx, mut block_state)| -> Box<dyn Future<Item = _, Error = _> + Send> {
+                if block_state.created_data_sources.is_empty() {
+                    // No new data sources, nothing to do.
+                    return Box::new(future::ok(Loop::Break((ctx, block_state))));
+                }
 
-            if triggers.len() == 1 {
-                info!(
-                    logger,
-                    "1 trigger found in this block for the new data sources"
-                );
-            } else if triggers.len() > 1 {
-                info!(
-                    logger,
-                    "{} triggers found in this block for the new data sources",
-                    triggers.len()
-                );
-            }
+                // Instantiate dynamic data sources, removing them from the block state.
+                let (data_sources, runtime_hosts) = match create_dynamic_data_sources(
+                    logger.clone(),
+                    &ctx.inputs,
+                    block_state.created_data_sources.drain(..),
+                ) {
+                    Ok(ok) => ok,
+                    Err(err) => return Box::new(future::err(err.into())),
+                };
+
+                // Reprocess the triggers from this block that match the new data sources
+
+                let block_with_calls = EthereumBlockWithCalls {
+                    ethereum_block: block.deref().clone(),
+                    calls: calls.clone(),
+                };
```

Comment:
This code is wrong afaict, if the new data source contains call handlers for new addresses the call won't be contained here. But this is the way it was before so I will open an issue for fixing this.

Reply:
Looked into this and found that the calls will not be included in the block in the case where the data sources at the beginning of the block did not require them. The calls will be complete when they are included though.